content/ActiveAdmin 2FA with OneLogin.md

```diff
@@ -0,0 +1,190 @@
+
+```ruby
+# app/controllers/admin_users/sessions_controller.rb
+
+module AdminUsers
+  class SessionsController < Devise::SessionsController
+    # As you are overwriting devise session controller you need this to allow to login with user & pass (dev mode)
+    prepend_before_action :require_no_authentication, only: [:new, :create]
+
+    layout 'active_admin_logged_out'
+    helper ::ActiveAdmin::ViewHelpers
+
+    def new
+      unless Rails.env.development? || Rails.env.test?
+        return redirect_to :new_saml_admin_user_session
+      end
+
+      super
+    end
+  end
+end
+```
+
+```ruby
+# config/initializers/active_admin_devise.rb
+module ActiveAdmin
+  module Devise
+    def self.controllers
+      {
+        sessions: "admin_users/sessions",
+        passwords: "active_admin/devise/passwords",
+        unlocks: "active_admin/devise/unlocks",
+        registrations: "active_admin/devise/registrations",
+        confirmations: "active_admin/devise/confirmations"
+      }
+    end
+
+    def self.controllers_for_filters
+      [
+        ::AdminUsers::SessionsController,
+        SessionsController,
+        PasswordsController,
+        UnlocksController,
+        RegistrationsController,
+        ConfirmationsController,
+      ]
+    end
+  end
+end
+```
+
+###
+**And that's all? wow!!!**
+
+Well.... not really, sorry for getting you excited ¯\\_(ツ)_/¯
+
+As you can see in the initializer there are some conditionals for development, that's because there is no dev out there that wants to do a 2FA every time they want to access ActiveAdmin locally.
+So we still need to add some code to conditionally enable 2FA depending on the environment, and a flag (because we don't want to block access to admin if something happens to OneLogin also)
+
+We need to check:
+  - Usual admin login page works in development
+  - Turning the feature off uses the old admin login and it works
+  - How logout works
+
+
+To perform the first two points I ended up adding the next lines to our overridden sessions_controller, seems that they are lost from super as you override it
+
+``` ruby
+# app/controllers/admin_users/sessions_controller.rb
+
+prepend_before_action :require_no_authentication, only: [:new, :create]
+prepend_before_action :allow_params_authentication!, only: :create
+prepend_before_action :verify_signed_out_user, only: :destroy
+prepend_before_action(only: [:create, :destroy]) { request.env["devise.skip_timeout"] = true }
+```
+
+**How logout works**
+
+At this point you have 3 options to logout:
+  - destroy your app session
+  - destroy saml session
+  - force logout from your third party auth partner
+
+For this specific case we needed to perform the first one only, because logging someone out of OneLogin means they will be logged out of a lot of apps, and that was not desired.
```

Comment:
```suggestion
For this specific case, we needed to perform the first one only, because logging someone out of OneLogin means they will be logged out of a lot of apps, and that was not desired.
```